### PR TITLE
[WIP] ENH: Add Array API compatibility infrastructure

### PR DESCRIPTION
**Pull request by mataton**
<html><body>
<html><head></head><body><h1>Array API Compatibility Decorator</h1>
<p>This PR introduces infrastructure for writing and testing array API compatible functions across multiple backends (NumPy, JAX, PyTorch, CuPy, Dask).</p>
<h2>Components</h2>
<p><strong><code>@array_api_compatible(backends, devices)</code></strong> — Decorator for production code. Marks a function as supporting specific backends/devices. It:</p>
<ul>
<li>Injects a compatibility table into the docstring's Notes section</li>
<li>Validates at runtime that the caller's array comes from a supported backend</li>
<li>Attaches <code>_array_api_backends</code> and <code>_array_api_devices</code> attributes to the function</li>
</ul>
<p><strong><code>@backends(*backend_names, cpu_only=False)</code></strong> — Decorator for test methods. Runs the test once per backend/device combination using <code>unittest.subTest()</code>. Backends are controlled via environment variables, so CI can test each backend in isolation.</p>
<p><strong><code>ArrayAPITestMixin</code></strong> — Mixin class providing test helpers:</p>
<ul>
<li><code>make_array(xp, device, data)</code> — Create an array on the specified backend/device</li>
<li><code>assert_close(actual, expected)</code> — Compare arrays (auto-converts GPU arrays to NumPy)</li>
<li><code>assert_type_preserved(result, xp, device)</code> — Verify the result stayed on the correct backend/device</li>
</ul>
<p><strong><code>ingest_array()</code></strong> — Utility (in <code>skbio.util._array</code>) that normalizes input arrays and extracts the array namespace, used internally by compatible functions.</p>
<h2>Environment Variables</h2>

Variable | Values | Default
-- | -- | --
SKBIO_ARRAY_API_BACKEND | numpy, jax, torch, cupy, all | numpy
SKBIO_DEVICE | cpu, cuda, gpu | cpu

<h2>CI Workflow</h2>
<p>The new <code>.github/workflows/array-api.yml</code> runs weekly tests against torch, jax, and cupy backends.</p>
<hr>
<h2>For Developers</h2>
<h3>Marking a function as array API compatible</h3>
<pre><code class="language-python">from skbio.util._decorator import array_api_compatible

@array_api_compatible(
    backends=["numpy", "cupy", "torch", "jax", "dask"],
    devices=["cpu", "gpu"],
)
def closure(mat, axis=-1, validate=True):
    ...
</code></pre>
<h3>Writing tests for array API functions</h3>
<pre><code class="language-python">from unittest import TestCase
from skbio.util._testing import ArrayAPITestMixin, backends

class TestClosure(TestCase, ArrayAPITestMixin):

    @backends("numpy", "jax", "torch", "cupy")
    def test_basic(self, xp, device):
        data = self.make_array(xp, device, [[2.0, 2.0, 6.0]])
        result = closure(data)
        
        self.assert_type_preserved(result, xp, device)
        self.assert_close(result, [[0.2, 0.2, 0.6]])
</code></pre>
<h3>Running tests against different backends</h3>
<pre><code class="language-bash"># Default (numpy only)
python -m skbio.test

# Specific backend
SKBIO_ARRAY_API_BACKEND=jax python -m skbio.test

# All installed backends
SKBIO_ARRAY_API_BACKEND=all python -m skbio.test

# GPU testing
SKBIO_ARRAY_API_BACKEND=torch SKBIO_DEVICE=cuda python -m skbio.test
</code></pre></body></html>
</body>
</html>
Please complete the following checklist:

* [x] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [ ] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [x] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [ ] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

File: skbio/util/_decorator.py [hunk 1]
```diff
@@ -18,10 +18,58 @@
...
     _renaming_note,
+    _array_api_compat_section,
+    _insert_into_notes_section,
 )
+from ._array import _get_namespace_from_args, _check_array_api_backend
 from ._warning import _warn_deprecated, _warn_renamed, _warn_param_renamed
 
 
+def array_api_compatible(backends, devices=None):
+    """Mark a function as compatible with the Python array API standard.
```
Comment (qiyunzhu) [inline skbio/util/_decorator.py]: The docstring should be prefixed with an `r`. Otherwise double ticks below won't be necessary and can be replaced with single ticks. Same in other cases.

File: skbio/util/_array.py [hunk 2]
```diff
@@ -85,6 +85,189 @@ def _get_array(arr: ArrayLike, /, *, to_numpy: bool = False) -> StdArray:
...
+    "numpy": aac.is_numpy_namespace,
+    "cupy": aac.is_cupy_namespace,
+    "torch": aac.is_torch_namespace,
+    "jax": aac.is_jax_namespace,
+    "dask": aac.is_dask_namespace,
+}
+
+
+def _get_namespace_from_args(args, kwargs):
+    """Return the array namespace if any argument is an array API object.
```
Comment (qiyunzhu) [inline skbio/util/_array.py]: According to [here](https://data-apis.org/array-api-compat/helper-functions.html#array_api_compat.is_array_api_obj), this term should be "array API compatible array object", which is too verbose. But "array API object" might be confusing. I would suggest just "array object".

File: skbio/util/_array.py [hunk 3]
```diff
@@ -85,6 +85,189 @@ def _get_array(arr: ArrayLike, /, *, to_numpy: bool = False) -> StdArray:
...
+    kwargs : dict
+        Keyword arguments.
+
+    Returns
+    -------
+    namespace or None
+        ``array_api_compat.array_namespace(*arrays)`` for all detected array
+        API objects, or ``None`` if none were found.
+
+    """
```
Comment (qiyunzhu) [inline skbio/util/_array.py:122]: `aac.array_namespace` raises a `TypeError` if arguments are different array types. This may be noted here for better understanding.

File: skbio/util/_array.py [hunk 4]
```diff
@@ -85,6 +85,189 @@ def _get_array(arr: ArrayLike, /, *, to_numpy: bool = False) -> StdArray:
...
+    Parameters
+    ----------
+    args : tuple
+        Positional arguments.
+    kwargs : dict
+        Keyword arguments.
+
+    Returns
+    -------
+    namespace or None
```
Comment (qiyunzhu) [inline skbio/util/_array.py:113]: This sentence is hard to understand. Can instead say "A uniform array namespace for all detected array objects, or `None` if no array object is found."

File: skbio/util/_decorator.py [hunk 5]
```diff
@@ -18,10 +18,58 @@
...
+    ``_array_api_backends`` and ``_array_api_devices``.
+
+    """
+
+    def decorator(func):
+        section = _array_api_compat_section(backends, devices)
+
+        @wraps(func)
+        def wrapper(*args, **kwargs):
+            xp = _get_namespace_from_args(args, kwargs)
```
Comment (qiyunzhu) [inline skbio/util/_decorator.py:59]: What if the user supplies Python lists? Will it trigger NumPy as a fallback backend? Or is this irrelevant here?

File: skbio/util/_docstring.py [hunk 6]
```diff
@@ -113,6 +113,148 @@ def _note_into_doc_param(note, doc, param):
...
+    col_headers = [d.upper() for d in all_devices]
+
+    # Build rows
+    rows = []
+    for b in backends:
+        display = _BACKEND_DISPLAY.get(b, b.capitalize())
+        supported = _BACKEND_DEVICES.get(b, ["cpu"])
+        if devices is not None:
+            supported = [d for d in devices if d in supported]
+        cells = ["\u2713" if d in supported else "n/a" for d in all_devices]
```
Comment (qiyunzhu) [inline skbio/util/_docstring.py:168]: Looks like there is no `cross` option here. What if a library + device combo is theoretically available but the specific function won't run correctly with that combo?

File: .github/workflows/array-api.yml [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
+      - name: Install scikit-bio + backend
+        run: |
+          pip install -e .
+          pip install ${{ matrix.backend }}
+      - name: Run tests
+        env:
+          SKBIO_ARRAY_API_BACKEND: ${{ matrix.backend }}
```
Comment (qiyunzhu) [inline .github/workflows/array-api.yml]: I think that `SKBIO_ARRAY_API_BACKEND` can be shortened as `SKBIO_ARRAY_BACKEND`. My rationale is that "API" is the interface, not the data structure.

Also a note: the word "backend" isn't used anywhere in the [Python array API standards](https://data-apis.org/array-api/latest/index.html) or the [array-api-compat documentation](https://data-apis.org/array-api-compat/index.html). Instead, "library" is what they use. I think we can keep "backend" because this term has already been used in scikit-bio to refer to different dataframe backends (Pandas, Polars, and NumPy).

File: .github/workflows/array-api.yml [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+            runs-on: ubuntu-latest
+          - backend: jax
+            runs-on: ubuntu-latest
+          - backend: cupy
+            runs-on: ubuntu-latest-gpu   # TODO: replace with your GPU runner label
+    steps:
+      - uses: actions/checkout@v4
+      - uses: actions/setup-python@v5
+        with:
+          python-version: "3.12"
```
Comment (qiyunzhu) [inline .github/workflows/array-api.yml]: Is there a reason to pin the code to Python 3.12?

File: skbio/util/_testing.py [hunk 9]
```diff
@@ -452,3 +456,231 @@ def pytestrunner():
...
+
+
+def _read_env():
+    """Read backend/device environment variables."""
+    backend = os.environ.get("SKBIO_ARRAY_API_BACKEND", "").strip()
+    device = os.environ.get("SKBIO_DEVICE", "").strip() or None
+    return backend, device
+
+
+def _get_available_backends():
```
Comment (qiyunzhu) [inline skbio/util/_testing.py]: Should it be `_get_array_backends`? Because there are dataframe backends too (Pandas, Polars and NumPy).

File: skbio/util/_testing.py [hunk 10]
```diff
@@ -452,3 +456,231 @@ def pytestrunner():
...
+        except RuntimeError:
+            pass
+        _backends["jax"] = (jnp, devices)
+    except ImportError:
+        pass
+
+    try:
+        import torch
+
+        torch.set_default_dtype(torch.float64)
```
Comment (qiyunzhu) [inline skbio/util/_testing.py]: What's the rationale of this line?

File: skbio/util/_testing.py [hunk 11]
```diff
@@ -452,3 +456,231 @@ def pytestrunner():
...
+    """Assert two arrays are element-wise exactly equal.
+
+    Converts both arrays to NumPy (handles GPU arrays), then delegates to
+    ``numpy.testing.assert_array_equal``.
+
+    """
+    npt.assert_array_equal(_to_numpy(actual), _to_numpy(desired))
+
+
+def backends(*backend_names, cpu_only=False):
```
Comment (qiyunzhu) [inline skbio/util/_testing.py]: Likewise, I think it should be `array_backends`.

File: skbio/util/_decorator.py [hunk 12]
```diff
@@ -18,10 +18,58 @@
...
+            if xp is not None:
+                _check_array_api_backend(xp, backends, func.__name__)
+            return func(*args, **kwargs)
+
+        wrapper.__doc__ = _insert_into_notes_section(section, func.__doc__ or "")
+        wrapper._array_api_backends = backends
+        wrapper._array_api_devices = devices
```
Comment (qiyunzhu) [inline skbio/util/_decorator.py]: There is a discrepancy in naming: For environment variables you have `SKBIO_ARRAY_API_BACKEND` and `SKBIO_DEVICE`. But here you have `_array_api_backends` and `_array_api_devices`. I would suggest using `_array_backends` and `_devices`, because "device" is not specific to arrays.

**Comment by qiyunzhu** [comment]
Per our offline discussion, I will merge this PR and move from here. There are some unaddressed issues, which we will fix in the near future. The fixations won't change the interface that is relevant to users and developers.

**Comment by Aaraviitkgp** [comment]
@qiyunzhu can I now use this to write some tests or should I wait ?

**Comment by qiyunzhu** [comment]
@Aaraviitkgp Yes! This should have unblocked your work. Thank you!
